clang/lib/Parse/ParseOpenACC.cpp

```diff
@@ -45,11 +47,36 @@ OpenACCDirectiveKind ParseOpenACCDirectiveKind(Parser &P) {
   P.ConsumeToken();
   std::string FirstTokSpelling = P.getPreprocessor().getSpelling(FirstTok);
 
-  OpenACCDirectiveKind DirKind = GetOpenACCDirectiveKind(FirstTokSpelling);
+  OpenACCDirectiveKind DirKind = getOpenACCDirectiveKind(FirstTokSpelling);
 
   if (DirKind == OpenACCDirectiveKind::Invalid)
     P.Diag(FirstTok, diag::err_acc_invalid_directive) << FirstTokSpelling;
 
+  // Combined Constructs allows parallel loop, serial loop, or kernels loop. Any
+  // other attempt at a combined construct will be diagnosed as an invalid
+  // clause.
+  Token SecondTok = P.getCurToken();
+  if (!SecondTok.isAnnotation() &&
+      P.getPreprocessor().getSpelling(SecondTok) == "loop") {
```

Comment:
Better to use `getOpenACCDirectiveKind` here to aavoid spreading constant strings across code.

Reply:
Hmm... while I see the benefit of that, I'm actually a bit concerned about the performance implications of doing what amounts to a dozen string comparisons in that case, vs 1 here.  Extracting this above has already increased the number of checks significantly (by checking for loop ALWAYS rather than just 1x), but it seems significantly worse to do multiples here.  WDYT?

Reply:
The better to add a function isOpenACCDirectiveKind(OpenACCDirectiveKind Kind, StringRef Tok)
```
switch (Kind) {
case OpenACCDirectiveKind::Loop:
  return Tok == "loop";
case OpenACCDirectiveKind::Parallel:
  return Tok == "parallel";
...
}
```

Reply:
Done.